{
 "metadata": {
  "name": "Step1- General Design, input, output"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Step1 - General Design, input, output"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Cross contamination of sample in NGS due to wrong attribution of index\n",
      "\n",
      "We experienced some problems of incorrect read attribution during Illumina read demultiplexing using simple sample indexing with CASAVA. This led to a cross contamination of reads from one sample to another, which can be problematic for low frequency variant detection or other specific usage of NGS. An experimental approach to reduce the number of incorrectly attributed cluster is to add a second sample index in p5 adapter. Bioinformatics can also provide an additional answer by taking into account the PHRED quality of index read(s). CASAVA is not able to manage quality of INDEX reads, but by modifying the sample sheet, it's possible to output non demultiplexed fastq files (see [convert-bcl-to-fastq](https://gist.github.com/brantfaircloth/3125885)). In this case, index read(s) will be output in separate fastq file(s) synchronized with read1 and read2. I want to design an Illumina paired end read demultiplexer to attribute each read pair to a given sample based on the quality of the index read.\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Quantificative bias for high depth sequencing.\n",
      "\n",
      "When the depth of coverage (DOC) is very high, PCR duplicates and independent sonication fragment cannot by differentiate. Given that the sequence composition and secondary structures can influence the efficiency of PCR and sequencing steps, quantifying sequences with heterogeneous DOC is inherently biased. To correct this bias, it is possible to add a stochastic sequence after the sample index (several N). This sequences also called molecular index will be included at the end of INDEX reads and can be subsequently used to differentiate PCR replicates by checking if read mapped at the same position have different molecular index. Using such molecular index required modification of bioinformatic tools, to extract the sequence of the molecular index during demultiplexing.  "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**To summarize, the demultiplexer need to handle dual sample indexing, including INDEX read quality, and should be able to extract one or 2 molecular index**.\n",
      "\n",
      "The following picture illustrate the most complex situation with 2 sample indexes and 2 molecular indexes.\n",
      "\n",
      "![Design](files/img/modified_illumina_adapters.png)"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Defining program IO\n",
      "\n",
      "### Input\n",
      "\n",
      "Without demultiplexing, CASAVA generates 1 file per sequencing reading:\n",
      "   \n",
      "* R1 = Forward insert\n",
      "* R2 = First index\n",
      "* R3 = Reverse insert\n",
      "* [R4 = Second index read] (if a second index is used)\n",
      "\n",
      "Reads in files are synchronized, so that for instance the 4rd read of R1 correspond to the same cluster than the 4rd read of R2, R3 and R4. Files are usually divided in ~ 600 mo chunks.\n",
      "I found file paths simpler to be indicated as CLI arguments and the other options in a configuration file that could be subsequently parsed by the program.\n",
      "Files path to fastq files will be required as CLI arguments in a specific order: \n",
      "\n",
      "chunk1-R1.fastq  chunk1-R2.fastq  chunk1-R3.fastq chunk1-R4.fastq  chunk2-R1.fastq  chunk2-R2.fastq  chunk2-R3.fastq  chunk2-R4.fastq ...\n",
      "\n",
      "The configuration file will contain values such as the sequence of index for each sample, the minimal phred quality of index... I usually write first a draft configuration file that can evolved during program dev to include new options.\n",
      "I know that my configuration file will be parse by the python module ConfigParser, so I wrote the following conf file compatible with its simple syntax. "
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "#########################################################################################\n",
      "#                                                                                       #\n",
      "#                            QUADE CONFIGURATION FILE                                   #\n",
      "#                                                                                       #\n",
      "#########################################################################################                                                    \n",
      "\n",
      "#   Values can by customized with users values, but the file template must remain unchanged, otherwise the program will not be able to load default values.\n",
      "#   Please follow the following recommendations :\n",
      "#   - File path should be indicated as absolute path preferably and should not contain blank spaces\n",
      "#   - Values identified with '**' in the descriptor are not recommended to be modified\n",
      "\n",
      "##########\n",
      "[general]\n",
      "##########\n",
      "\n",
      "# Path where to create a folder in which all files generated during the program will be stored. Can be leaved empty if you want to work in the current directory (string)\n",
      "outdir : ./my_folder\n",
      "\n",
      "##########\n",
      "[quality]\n",
      "##########\n",
      "\n",
      "# Minimal mean quality of the index to consider a read pair valid. Leave empty if not requested. (INTEGER)\n",
      "mean_qual : 30\n",
      "\n",
      "# Minimal quality for one base of the index to consider a read pair valid. Leave empty if not requested. (INTEGER)\n",
      "minimal_qual : 20\n",
      "\n",
      "# Quality format associated with reads autorized values are : 'fastq-sanger' 'fastq-solexa' or \n",
      "# 'fastq-illumina'. See Biopython QualityIO documentation for more details. (STRING)\n",
      "quality_scale : fastq_sanger\n",
      "\n",
      "##########\n",
      "[index]\n",
      "##########\n",
      "\n",
      "# Definition of sample and molecular indexes positions within in index read using 0 base coordinates. Leave empty if the index is not in use.\n",
      "\n",
      "# Sample index 1 start and end positions (INTEGER)\n",
      "sample1_start : 0\n",
      "sample1_end : 5\n",
      "\n",
      "# Sample index 2 start and end positions [if double indexing] (INTEGER)\n",
      "sample2_start : 0\n",
      "sample2_end : 5\n",
      "\n",
      "# Molecular index 1 if stochastic index was added after sample index 1 (INTEGER)\n",
      "molecular1_start :\n",
      "molecular1_end :\n",
      "\n",
      "# Molecular index 1 if stochastic index was added after sample index 2 (INTEGER)\n",
      "molecular1_start :\n",
      "molecular1_end :\n",
      "\n",
      "###############################\n",
      "#\u00a0   SAMPLE DEFINITIONS\u00a0      #\n",
      "###############################\n",
      "\n",
      "# It is possible to include as many independant sample as required by duplicating a entire sample section and incrementing the id number starting at 1.\n",
      "\n",
      "[sample1] \n",
      "\n",
      "# basename that will be used to prefix the read files (STRING)\n",
      "name : S1\n",
      "\n",
      "# Index sequences associated with the sample. Index 2 is used only in case of double indexing (STRING)\n",
      "index1_seq : ATCGT\n",
      "index2_seq : TTGTG\n",
      "\n",
      "[sample2]\n",
      "\n",
      "# basename that will be used to prefix the read files (STRING)\n",
      "name : S2\n",
      "\n",
      "# Index sequences associated with the sample. Index 2 is used only in case of double indexing (STRING)\n",
      "index1_seq : ATCCT\n",
      "index2_seq : TTGGG\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Output\n",
      "\n",
      "I want to obtain a couple of fastq.gz files for each sample, containing reads corresponding to the sample index specified in the configuration file. The name of each read needs to contain sequences from the sample and molecular index, in order to be easily accessible for post-treatments.  \n",
      "\n",
      "If index quality is not good enough or if the index (or index combination) does not match any sample, these reads should be stored in unidentified fastq files.\n",
      "\n",
      "Finally, I also want the program to generate a simple text report summarizing what was done during the analysis.\n",
      "\n",
      "![](files/img/quade-io.png)"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Parsing the configuration file\n",
      "\n",
      "Generally, I like to use python class to handle my main instructions. Configuration file and CLI argument parsing, value verifications and variable storing can be elegantly integrated into the initialization of the class (__init__).\n",
      "\n",
      "Top level running instruction can then be written in the magic call method to make your object directly callable.\n",
      "\n",
      "Here is a first version of the main class adapted for interactive interpreters"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Imports\n",
      "import ConfigParser\n",
      "\n",
      "# Definition of the main class\n",
      "class Main(object):\n",
      "    \n",
      "    # Class init method \n",
      "    def __init__ (self, conf_file): # conf_file arg is only for interactive interpreter. Will be remove in final program\n",
      "        \n",
      "        # Define a configuration file parser object\n",
      "        self.conf = ConfigParser.RawConfigParser(allow_no_value=True)\n",
      "        \n",
      "        # Indicate the path to the configuration file\n",
      "        self.conf.read(conf_file)\n",
      "        \n",
      "        # In a try block due to possible errors while reading the file \n",
      "        try:\n",
      "            #### PARSE CONF FILE ####\n",
      "            \n",
      "            # Get Values from the required section in the appropriate type \n",
      "            self.outdir = self.conf.get(\"general\", \"outdir\") # for string values\n",
      "            self.mean_qual = self.conf.getint(\"quality\", \"mean_qual\") # for integer values\n",
      "            self.minimal_qual = self.conf.getint(\"quality\", \"minimal_qual\")\n",
      "            self.quality_scale = self.conf.get(\"quality\", \"quality_scale\")\n",
      "            \n",
      "            # for each index verify its existence and store start and end positions if needed \n",
      "            self.index1 = self.conf.getboolean(\"index\", \"index1\")  # for boolean values\n",
      "            if self.index1:\n",
      "                self.index1_start = self.conf.getint(\"index\", \"index1_start\")\n",
      "                self.index1_end = self.conf.getint(\"index\", \"index1_end\")\n",
      "            self.index2 = self.conf.getboolean(\"index\", \"index2\")\n",
      "            if self.index2:\n",
      "                self.index2_start = self.conf.getint(\"index\", \"index2_start\")\n",
      "                self.index2_end = self.conf.getint(\"index\", \"index2_end\")\n",
      "            self.molecular1 = self.conf.getboolean(\"index\", \"molecular1\")\n",
      "            if self.molecular1:\n",
      "                self.molecular1_start = self.conf.getint(\"index\", \"molecular1_start\")\n",
      "                self.molecular1_end = self.conf.getint(\"index\", \"molecular1_end\")\n",
      "            self.molecular2 = self.conf.getboolean(\"index\", \"molecular1\")\n",
      "            if self.molecular2:\n",
      "                self.molecular2_start = self.conf.getint(\"index\", \"molecular2_start\")\n",
      "                self.molecular2_end = self.conf.getint(\"index\", \"molecular2_end\")\n",
      "            \n",
      "            # Sample section are a special case since the number of sample section is variable\n",
      "            # Only section with a name starting by \"sample\" will be parsed\n",
      "            self.sample_list = []\n",
      "            for sample in [i for i in parser.sections() if i.startswith(\"sample\")]:\n",
      "                self.sample_list.append({\n",
      "                    'name'   : self.conf.get(sample, \"name\"),\n",
      "                    'index1_seq' :  self.conf.get(sample, \"index1_seq\"),\n",
      "                    'index2_seq' :  self.conf.get(sample, \"index2_seq\")})\n",
      "            \n",
      "            #### TEST VALUES ####\n",
      "            \n",
      "            # Now all variable can be tested and an assertion error will be raised in case of invalid value\n",
      "            self.outdir = self.outdir if self.outdir else \"./\"\n",
      "            self.quality_scale = self.conf.get(\"quality\", \"quality_scale\")\n",
      "            \n",
      "            assert self.mean_qual >= 0 and self.mean_qual <= 40, \"Autorized values for mean_qual : 0 to 40\"\n",
      "            assert self.mean_qual >= 0 and self.mean_qual <= 40, \"Autorized values for minimal_qual : 0 to 40\"\n",
      "            if self.index1:\n",
      "                assert self.index1_start >= 0, \"Autorized values for index1_start : >= 0\"\n",
      "                assert self.index1_end >= self.index1_start, \"Autorized values for index1_end : > index1_start\"\n",
      "            if self.index2:\n",
      "                assert self.index2_start >= 0, \"Autorized values for index2_start : >= 0\"\n",
      "                assert self.index2_end >= self.index2_start, \"Autorized values for index2_end : > index2_start\"\n",
      "            if self.molecular1:\n",
      "                assert self.molecular1_start >= 0, \"Autorized values for molecular1_start : >= 0\"\n",
      "                assert self.molecular1_end >= self.molecular1_start, \"Autorized values for molecular_end : > molecular1_start\"\n",
      "            if self.molecular2:\n",
      "                assert self.molecular2_start >= 0, \"Autorized values for molecular2_start : >= 0\"\n",
      "                assert self.molecular2_end >= self.molecular2_start, \"Autorized values for molecular2_end : > molecular2_start\"\n",
      "                \n",
      "            \n",
      "        # Handle the possible exception by printing simple message for users.\n",
      "        except ConfigParser.NoOptionError as E:\n",
      "            print (E)\n",
      "            print (\"An option is missing in the configuration file\")\n",
      "            print (\"Please report to the descriptions in the configuration file\\n\")\n",
      "            exit()\n",
      "        except ConfigParser.NoSectionError as E:\n",
      "            print (E)\n",
      "            print (\"An section is missing in the configuration file\")\n",
      "            print (\"Please report to the descriptions in the configuration file\\n\")\n",
      "            exit()\n",
      "        except (ValueError, AssertionError) as E:\n",
      "            print (E)\n",
      "            print (\"One of the value in the configuration file is not in the correct format\")\n",
      "            print (\"Please report to the descriptions in the configuration file\\n\")\n",
      "            exit()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": "*"
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "main = Main(\"../Conf.txt\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": "*"
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "main"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": "*"
    }
   ],
   "metadata": {}
  }
 ]
}